ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound.hs

```diff
@@ -0,0 +1,527 @@
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Sequence.Strict (StrictSeq)
+import Data.Sequence.Strict qualified as Seq
+import Data.Set qualified as Set
+import Data.Word (Word16)
+import GHC.Generics (Generic)
+import Network.TypedProtocol.Core (N, Nat (..), natToInt)
+import NoThunks.Class (NoThunks (..), unsafeNoThunks)
+import Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.ObjectPool.API
+import Ouroboros.Network.NodeToNode.Version (NodeToNodeVersion)
+import Ouroboros.Network.Protocol.ObjectDiffusion.Inbound
+import Ouroboros.Network.Protocol.ObjectDiffusion.Type
+
+-- Note: This module is almost a perfect copy of TxSubmission inbound side.
+
+data ProcessedObjectCount objectId object = ProcessedObjectCount
+  { pocAcceptedObjectsCount :: Int
+  -- ^ Just accepted this many objects.
+  , pocRejectedObjectsCount :: Int
+  -- ^ Just rejected this many objects.
+  }
+  deriving (Eq, Show)
+
+data TraceObjectDiffusionInbound objectId object
+  = -- | Number of objects just about to be inserted.
+    TraceObjectDiffusionCollected Int
+  | -- | Just processed object pass/fail breakdown.
+    TraceObjectDiffusionProcessed (ProcessedObjectCount objectId object)
+  | -- | Inbound received 'MsgDone'
+    TraceObjectInboundTerminated
+  | TraceObjectInboundCanRequestMoreObjects Int
+  | TraceObjectInboundCannotRequestMoreObjects Int
+  deriving (Eq, Show)
+
+data ObjectDiffusionInboundError
+  = ProtocolErrorObjectNotRequested
+  | ProtocolErrorObjectIdsNotRequested
+  deriving Show
+
+instance Exception ObjectDiffusionInboundError where
+  displayException ProtocolErrorObjectNotRequested =
+    "The peer replied with a object we did not ask for."
+  displayException ProtocolErrorObjectIdsNotRequested =
+    "The peer replied with more objectIds than we asked for."
+
+-- | Information maintained internally in the 'objectDiffusionInbound' inbound
+-- implementation.
+data InboundState objectId object = InboundState
+  { reqObjectIdsInFlightCount :: !Word16
+  -- ^ The number of object identifiers that we have requested but
+  -- which have not yet been replied to. We need to track this it keep
+  -- our requests within the limit on the number of unacknowledged objectIds.
+  , unacknowledgedObjectIds :: !(StrictSeq objectId)
+  -- ^ Those objects (by their identifier) that the client has told
+  -- us about, and which we have not yet acknowledged. This is kept in
+  -- the order in which the client gave them to us. This is the same order
+  -- in which we submit them to the objectPool (or for this example, the final
+  -- result order). It is also the order we acknowledge in.
+  , availableObjectids :: !(Map objectId SizeInBytes)
+  -- ^ Those objects (by their identifier) that we can request. These
+  -- are a subset of the 'unacknowledgedObjectIds' that we have not yet
+  -- requested. This is not ordered to illustrate the fact that we can
+  -- request objects out of order. We also remember the size.
```

Comment:
Side note: It is plausible that our objects will have (almost) uniform size (certain for votes, *probably* for certificates depending on the cryptographic scheme), in which case tracking the byte size is unnecessary (in contrast to txs).

Reply:
Following our call from this morning, we decided to remove the SizeInBytes, so I'm closing this as resolved :)